src/charm.py

```diff
@@ -20,6 +20,39 @@
 DEFAULT_METRICS_ENDPOINT_PORT = 80
 
 
+def _validated_address(address):
+    # split host and port parts
+    num_colons = address.count(":")
+    if num_colons > 1:
+        logger.error("No ':' in target: %s", address)
+        return ""
+
+    host, port = address.split(":") if num_colons else (address, DEFAULT_METRICS_ENDPOINT_PORT)
+
+    # validate port
+    try:
+        port = int(port)
+    except ValueError:
+        logger.error("Invalid port for target: %s", port)
+        return ""
+
+    if port < 0 or port > 2 ** 16 - 1:
+        logger.error("Invalid port range for target: %s", port)
+        return ""
+
+    # validate host
+    match = re.search(
+        r"^(?:(?:(?:(?:[a-zA-Z0-9][-a-zA-Z0-9]*)?[a-zA-Z0-9])[.])*(?:[a-zA-Z][-a-zA-Z0-9]*[a-zA-Z0-9]|[a-zA-Z])[.]?)$",
```

Comment:
This regexp does not work for plain IP addresses.
Why not use something as simple as
```python
>>> from urllib.parse import urlparse
>>> urlparse('http://localhost:9090/metrics')
ParseResult(scheme='http', netloc='localhost:9090', path='/metrics', params='', query='', fragment='')
```
I'd require the user to provide a scheme, and drop it later. Also, the `/metrics` path does not need to be its own config param.